File: reverse_vowels_of_a_string#345/solution_1.js
/**
 * @param {string} s
 * @return {string}
 */
var reverseVowels = function(s) {
  const isVowels = char => { return ['a', 'e', 'i', 'o', 'u'].includes(char.toLowerCase()); };
  const swap = (arr, a, b) => { let t = arr[a]; arr[a] = arr[b]; arr[b] = t; };
  const chars = Array.from(s);
  let [p1, p2] = [0, chars.length-1];

  while (p1 < p2) {
    if (isVowels(chars[p1]) && isVowels(chars[p2])) {
       swap(chars, p1, p2);
       p1++
       p2--;
    } else {
        if (!isVowels(chars[p1]) && isVowels(chars[p2])) {
           p1++;
        }
        else if (isVowels(chars[p1]) && !isVowels(chars[p2])) {
           p2--;
        }
        else {
           p1++;
           p2--;
        }
    }
  }
  return chars.join("");
};

File: reverse_vowels_of_a_string#345/solution_1.rb
# @param {String} s
# @return {String}


def reverse_vowels(s)
  s = s.chars
  is_vowels = ->(c) { %w[a e i o u A E I O U].include?(c) }

  p1, p2 = [0, s.length-1]
  while p1 < p2
    if is_vowels.(s[p1]) && is_vowels.(s[p2])
      s[p1], s[p2] = s[p2], s[p1]
      p1 += 1
      p2 -= 1
      next
    end
    if is_vowels.(s[p1]) && !is_vowels.(s[p2])
      p2 -= 1
    else
      p1 += 1
    end
  end
  s.join
end

